File: kotlin/src/main/kotlin/com/royaltics/errortracker/core/Transport.kt
package com.royaltics.errortracker.core

import com.royaltics.errortracker.types.ClientConfig
import com.royaltics.errortracker.types.TransportPayload
import kotlinx.coroutines.delay
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import java.util.concurrent.TimeUnit

class Transport(private val config: ClientConfig) {
    private val json = Json { ignoreUnknownKeys = true }
    
    private val client = OkHttpClient.Builder()
        .connectTimeout(config.timeout, TimeUnit.MILLISECONDS)
        .writeTimeout(config.timeout, TimeUnit.MILLISECONDS)
        .readTimeout(config.timeout, TimeUnit.MILLISECONDS)
        .build()

    suspend fun send(compressedEvent: String) {
        val payload = TransportPayload(
            event = compressedEvent,
            licenseId = config.licenseId,
            licenseName = config.licenseName,
            licenseDevice = config.licenseDevice
        )

        var lastError: Exception? = null

        repeat(config.maxRetries + 1) { attempt ->
            try {
                makeRequest(payload)
                return
            } catch (e: Exception) {
                lastError = e
                if (attempt < config.maxRetries) {
                    delay(calculateBackoff(attempt))
                }
            }
        }

        throw lastError ?: IOException("Transport failed with unknown error")
    }

    private fun makeRequest(payload: TransportPayload) {
        val jsonPayload = json.encodeToString(payload)
        val mediaType = "application/json; charset=utf-8".toMediaType()
        val requestBody = jsonPayload.toRequestBody(mediaType)

        val requestBuilder = Request.Builder()
            .url(config.webhookUrl)
            .post(requestBody)
            .addHeader("Content-Type", "application/json")
            .addHeader("User-Agent", "Royaltics-ErrorTracker-Kotlin/1.0")

        config.headers.forEach { (key, value) ->
            requestBuilder.addHeader(key, value)
        }

        val request = requestBuilder.build()

        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) {
                throw IOException("HTTP ${response.code}: ${response.message}")
            }
        }
    }

    private fun calculateBackoff(attempt: Int): Long {
        val baseDelay = 1000L
        val maxDelay = 30000L
        return minOf(baseDelay * (1L shl attempt), maxDelay)
    }
}

File: flutter/lib/src/core/transport.dart
import 'dart:async';
import 'dart:convert';
import 'dart:math';
import 'package:http/http.dart' as http;
import '../types/client_config.dart';
import '../types/event_issue.dart';

class Transport {
  final ClientConfig config;
  final http.Client _client = http.Client();

  Transport(this.config);

  Future<void> send(String compressedEvent) async {
    final payload = TransportPayload(
      event: compressedEvent,
      licenseId: config.licenseId,
      licenseName: config.licenseName,
      licenseDevice: config.licenseDevice,
    );

    Exception? lastError;

    for (int attempt = 0; attempt <= config.maxRetries; attempt++) {
      try {
        await _makeRequest(payload);
        return;
      } catch (e) {
        lastError = Exception(e.toString());
        if (attempt < config.maxRetries) {
          await Future.delayed(_calculateBackoff(attempt));
        }
      }
    }

    throw lastError ?? Exception("Transport failed with unknown error");
  }

  Future<void> _makeRequest(TransportPayload payload) async {
    final jsonPayload = jsonEncode(payload.toJson());
    
    // We cannot set strict timeouts on per-request basis with http.Client easily without wrapping 
    // but we can assume the server handles it or we use .timeout()
    // However, clean http usage:
    
    final uri = Uri.parse(config.webhookUrl);
    final response = await _client.post(
      uri,
      headers: {
        'Content-Type': 'application/json',
        'User-Agent': 'Royaltics-ErrorTracker-Flutter/1.0',
        ...config.headers,
      },
      body: jsonPayload,
    ).timeout(Duration(milliseconds: config.timeout));

    if (response.statusCode < 200 || response.statusCode >= 300) {
      throw Exception("HTTP ${response.statusCode}: ${response.body}");
    }
  }

  Duration _calculateBackoff(int attempt) {
    const baseDelay = 1000;
    const maxDelay = 30000;
    final delayMs = min(baseDelay * (1 << attempt), maxDelay);
    return Duration(milliseconds: delayMs);
  }
}
